Expand on the "Intermediary" CNAME use-case in more detail (#86)
ietf-wg-dnsop/draft-ietf-dnsop-domain-verification-techniques: draft-ietf-dnsop-domain-verification-techniques.md

          - ins: Atlassian
        target: https://support.atlassian.com/user-management/docs/verify-a-domain-to-manage-accounts/#Verify-over-DNS

    SUBDOMAIN-TAKEOVER:

>         title: "Subdomain takovers"

Reviewer: ```suggestion
        title: "Subdomain Takeovers"
```

### Random Token {#random-token}

A unique token used in the challenge. It should be a random value issued to the user by the provider with the following properties:

1. MUST have at least 128 bits of entropy.
2. base64url ({{!RFC4648, Section 5}}) encoded or base16 ({{!RFC4648, Section 8}}) encoded.

See {{RFC4086}} for additional information on randomness requirements.

> Hexadecimal base16 encoding is recommended when the random token would exist in a DNS label such as in a CNAME target.

Reviewer: should this be normative language i.e. RECOMMENDED?

Reviewer: Also I think we should provide some rationale here, even if brief.

Reviewer 2: I guess it depends on whether the CNAME can resolve to a hostname, which inherits the LDH (letters digits hyphens) restriction. In that case, base64 alphabet violates the hostname rules, but base16 does not.

So, it depends on what the application provider uses the CNAME to point to.

Reviewer 2: But, lets's see if Erik has any additional rationale or justification for the use of this alphabet.

Author: ```suggestion
Hexadecimal base16 encoding is RECOMMENDED when the random token would exist in a DNS label such as in a CNAME target.  This is because base64 relies mixed case (and DNS is case-insensitive as clarified in {{RFC4343}}) and because some base64 characters ("/", "+", and "=") may not be permitted by implementations that limit allowed characters to those allowed in hostnames.
```

How is this?  We'd also need to add RFC4343 to the informational references.

Reviewer 2: Ok, yes, I think I can see a stronger rationale for such restrictions then. Technically, a domain name (as long as it isn't a hostname) permits arbitrary 8-bit characters. But yeah, provisioning systems may be likely to limit the character set, so perhaps a more restricted alphabet for domain name labels here will help avoid those problems.

Author: There are mechanisms such as case randomization in queries that assume a lack of case-sensitivity and are also deployed to mitigate Kaminsky attacks (https://lists.dns-oarc.net/pipermail/dns-operations/2023-January/021947.html and https://datatracker.ietf.org/doc/html/draft-vixie-dnsext-dns0x20-00).  Between RFC4343 and these trying to rely on mixed-case in labels seems like asking for trouble.  (There's an inherent conflict between the "permits arbitrary 8-bit characters" and the "case insensitive" of rfc4343.)

It may make sense to merge this overall PR #86 and then refine the details of the recommended encoding in a separate PR?  (I'm starting to envision merge conflicts with overlaps against this.)

> Hexadecimal base16 encoding is recommended when the random token would exist in a DNS label such as in a CNAME target.

Author: ```suggestion
Hexadecimal base16 encoding is RECOMMENDED when the random token would exist in a DNS label such as in a CNAME target.  This is because base64 relies mixed case (and DNS is case-insensitive as clarified in {{RFC4343}}) and because some base64 characters ("/", "+", and "=") may not be permitted by implementations that limit allowed characters to those allowed in hostnames.
```

How is this?  We'd also need to add RFC4343 to the informational references.

Reviewer: sounds good!

The user SHOULD de-provision the resource record provisioned for DNS-based domain control validation once it is no longer required.

## CNAME Records

CNAME records MAY be used instead of TXT records, either for Delegated Domain Control Validation ({{delegated}}) or where specified by providers to support users who are unable to create TXT records.

> A provider supporting CNAME records MUST specify the use of a label (e.g., `_foo-challenge`) as a CNAME MUST NOT be placed at the same domain name that is being validated. This is for the same reason already cited in {{pitfalls}}. CNAME records cannot co-exist with other data, and there may already be other record types that exist at the domain name. Instead, as with the TXT record recommendation, a provider specific label should be added as a subdomain of the domain to be verified. This ensures that the CNAME does not collide with other record types.

Reviewer 2: I think "MUST" specify the use of an application specific label is a good idea, and conforms to our general recommendations for TXT too. However, we should note that there are other providers that use CNAME and don't do this. For example Amazon Certificate Manger (ACM). to quote the example we cite in the appendix:

_<random-token1>.example.com.   IN   CNAME _<random-token2>.acm-validations.aws.

Here, their use of a random token is one way they avoid the collisions with other records at the same domain name.

Author: ```suggestion
A provider supporting CNAME records MUST specify the use of a label (e.g., "\_foo-<token>" or even the less-recommended "\_<token>") as a CNAME MUST NOT be placed at the same domain name that is being validated. This is for the same reason already cited in {{pitfalls}}. CNAME records cannot co-exist with other data, and there may already be other record types that exist at the domain name. Instead, as with the TXT record recommendation, a provider specific label should be added as a subdomain of the domain to be verified. This ensures that the CNAME does not collide with other record types.
```

How about this clarification?

    _foo-challenge.example.com.  IN   CNAME <random-token>.dcv.provider.example.

### Delegated Domain Control Validation {#delegated}

Separately, CNAME records also enable delegated domain control validation, which lets the user delegate the domain control validation process for their domain to an intermediary without having to hand over full DNS access. The intermediary gives the user a CNAME record to add for the domain and provider being validated that points to the intermediary's DNS, where the actual validation TXT record is placed. The record name and base16-encoded random tokens are generated as in {{format}}. For example:

    _foo-challenge.example.com.  IN   CNAME  "<intermediary-random-token>.dcv.intermediary.example."

The intermediary then adds the actual validation record in a domain they control:

    <intermediary-random-token>.dcv.intermediary.example. TXT "<provider-random-token>"

Such a setup is especially useful when the provider wants to periodically re-issue the challenge. CNAMEs allow automating the renewal process by letting the intermediary place the random token in their DNS instead of needing continuous write access to the user's DNS.

> Importantly, the CNAME record target also contains a random token issued by the intermediary to the user (preferably over a secure channel) which proves to the intermediary that example.com is controlled by the user. The intermediary must keep an association of users and domain names to the associated intermediary-random-tokens. Without a linkage validated by the intermediary during provisioning and renewal there is the risk that an attacker could leverage a "dangling CNAME" to perform a "Subdomain takover" attack ({{SUBDOMAIN-TAKEOVER}}).

Reviewer: ```suggestion
Importantly, the CNAME record target also contains a random token issued by the intermediary to the user (preferably over a secure channel) which proves to the intermediary that example.com is controlled by the user. The intermediary must keep an association of users and domain names to the associated intermediary-random-tokens. Without a linkage validated by the intermediary during provisioning and renewal there is the risk that an attacker could leverage a "dangling CNAME" to perform a "subdomain takover" attack ({{SUBDOMAIN-TAKEOVER}}).
```
